p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,516 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/connmgr"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/pnet"
+	"github.com/libp2p/go-libp2p/core/sec"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/encoding"
+
+	logging "github.com/ipfs/go-log/v2"
+	ma "github.com/multiformats/go-multiaddr"
+	mafmt "github.com/multiformats/go-multiaddr-fmt"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multihash"
+	pionlogger "github.com/pion/logging"
+
+	"github.com/pion/webrtc/v3"
+)
+
+var log = logging.Logger("webrtc-transport")
+
+var dialMatcher = mafmt.And(mafmt.UDP, mafmt.Base(ma.P_WEBRTC), mafmt.Base(ma.P_CERTHASH))
+
+const (
+	// handshakeChannelNegotiated is used to specify that the
+	// handshake data channel does not need negotiation via DCEP.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values.
+	handshakeChannelNegotiated = true
+	// handshakeChannelId is the agreed ID for the handshake data
+	// channel. A constant is used since the `DataChannelInit` struct takes
+	// references instead of values. We specify the type here as this
+	// value is only ever copied and passed by reference
+	handshakeChannelId = uint16(0)
```

Comment:
Nit: naming conventions
```suggestion
	handshakeChannelID = uint16(0)
```